monitoring/check_collectd_mlab.py

```diff
@@ -0,0 +1,461 @@
+
+  # Is utility slice quota ok?
+  exit_code = run_collectd_nagios(
+      'utility.mlab.' + HOSTNAME, 'storage/vs_quota_bytes-quota',
+      'used', _mb_to_bytes(8000), _mb_to_bytes(9000))
+  if exit_code != 0:
+    raise NagiosStateError('Storage quota usage is too high',
+                           status_code=exit_code)
+
+  # Is collectd cpu usage too high?
+  exit_code = run_collectd_nagios(
+      HOSTNAME, 'meta-collectd/process_cpu-system',
+      'value', '0:5', '0:10')
+  if exit_code != 0:
+    raise NagiosStateError('Collectd CPU usage is too high',
+                           status_code=exit_code)
+
+  # Actual memory usage should be pretty small, though vm size may be larger.
+  exit_code = run_collectd_nagios(
+      HOSTNAME, 'meta-collectd/process_memory-rss',
+      'value', _mb_to_bytes(10), _mb_to_bytes(15))
+  if exit_code != 0:
+    raise NagiosStateError('Collectd RSS memory usage is too high',
+                           status_code=exit_code)
+
+  # Is meta timer too high?
+  exit_code = run_collectd_nagios(
+      HOSTNAME, 'meta/timer-read',
+      'value', '0:6', '0:100')
+  if exit_code != 0:
+    raise NagiosStateError('Collectd mlab plugin taking too long to run',
+                           status_code=exit_code)
+
+
+def assert_disk_last_sync_time():
+  """Asserts the last sync time for vserver disk limits."""
+  # TODO: When was vserver quota last sync'd?
+  # TODO: How do we want to check this?
+  pass
+
+
+def check_collectd():
+  """Checks environment for signs of normal collectd-mlab behavior.
+  
+  Returns:
+    Tuple of (int, str), corresponding to the (nagios_state, error_message).
+  """
+  t_start = time.time()
+
+  # Check all critical conditions first.
+  try:
+    assert_collectd_installed()
+    assert_collectd_responds()
+    assert_collectd_vsys_setup()
+    assert_disk_last_sync_time()
+  except CriticalError as err:
+    return (STATE_CRITICAL, str(err))
+
+  # Since the above establishes that collectd is working, now check collectd
+  # resource usage thresholds *from* collectd.
+  try:
+    assert_collectd_nagios_levels()
+  except NagiosStateError as err:
+    return (err.status_code, str(err))
+
+  # We've made it this far, so everything appears to be ok.
+  msg = 'Blue skies smiling at me: %0.2f sec' % (time.time() - t_start)
+  return (STATE_OK, msg)
+
+
+def init_alarm(timeout):
+  """Assigns a SIGALARM handler for timeout seconds."""
+  def handler(unused_signum, unused_frame):
+    """Raise a TimeoutError when called"""
+    raise TimeoutError('Timeout after %s' % timeout)
+  signal.signal(signal.SIGALRM, handler)
+  signal.alarm(timeout)
+
+
+def cancel_alarm():
+  """Cancels pending alarm."""
+  signal.alarm(0)
+
+
+def main():
+  init_alarm(DEFAULT_TIMEOUT)
+
+  try:
+    (status_code, message) = check_collectd()
+  except Exception as err:  # pylint: disable=W0703
+    status_code = STATE_UNKNOWN
+    message = str(err)
+
+  cancel_alarm()
```

Comment:
(optional) Given my perennial love for context handlers, I suggest doing this in a context handler.